common/achievement/index.ts

```diff
@@ -0,0 +1,123 @@
+
+    const actionEvent: ActionEvent<ID> = {
+        actionId,
+        at: new Date(),
+        user: user,
+        context,
+    };
+    await trackEvent(actionEvent);
+
+    for (const [, group] of templatesByGroups) {
+        let achievementToCheck: AchievementToCheck;
+        for (const template of group) {
+            const userAchievement = await getOrCreateUserAchievement(template, user.userID, context);
+            if (userAchievement.achievedAt === null || userAchievement.recordValue) {
+                achievementToCheck = userAchievement;
+                break;
+            }
+        }
+        if (achievementToCheck) {
+            await checkUserAchievement(achievementToCheck as UserAchievementTemplate, actionEvent);
+        }
+    }
+}
+
+async function trackEvent<ID extends ActionID>(event: ActionEvent<ID>) {
+    const metricsForEvent = getMetricsByAction(event.actionId);
+
+    if (!metricsForEvent) {
+        logger.debug(`Can't track event, because no metrics found for action '${event.actionId}'`);
+        return;
+    }
+
+    for (const metric of metricsForEvent) {
+        const formula = metric.formula;
+        const value = formula(event.context);
+
+        await prisma.achievement_event.create({
+            data: {
+                metric: metric.metricName,
+                value: value,
+                action: event.actionId,
+                userId: event.user.userID,
+                relation: event.context.relation ?? '',
+            },
+        });
+    }
+
+    return true;
+}
+
+async function checkUserAchievement<ID extends ActionID>(userAchievement: UserAchievementTemplate, event: ActionEvent<ID>) {
+    const evaluationResult = await isAchievementConditionMet(userAchievement);
+
+    if (evaluationResult.conditionIsMet) {
+        const conditionDataAggregations = userAchievement?.template.conditionDataAggregations as ConditionDataAggregations;
+        const dataAggregationKey = Object.keys(conditionDataAggregations)[0];
+        const evaluationResultValue =
+            typeof evaluationResult.resultObject[dataAggregationKey] === 'number' ? Number(evaluationResult.resultObject[dataAggregationKey]) : null;
+        const awardedAchievement = await rewardUser(evaluationResultValue, userAchievement, event);
+        const userAchievementContext: UserAchievementContext = {};
+        await createAchievement(awardedAchievement.template, userAchievement.userId, userAchievementContext);
+    } else {
+        await prisma.user_achievement.update({
+            where: { id: userAchievement.id },
+            data: { achievedAt: null, isSeen: false },
+        });
+    }
+}
+
+async function isAchievementConditionMet(achievement: UserAchievementTemplate) {
+    const {
+        recordValue,
+        template: { condition, conditionDataAggregations, metrics },
+    } = achievement;
+    if (!condition) {
+        return;
+    }
```

Comment:
We should log an error in this case. Would be quite bad to have active achievements without condition.